server/pulp/server/db/model/__init__.py

```diff
@@ -307,3 +312,132 @@ def post_save(cls, sender, document, **kwargs):
 
 
 signals.post_save.connect(TaskStatus.post_save, sender=TaskStatus)
+
+
+class ContentUnit(Document):
+    """
+    The base class for all content units.
+
+    All unit inheriting from this class should override the unit_type_id and _ns to ensure
+    that they are populated properly
+
+
+    :ivar id: string representation of content unit id
+    :type id: mongoengine.StringField
+    :ivar last_updated: integer representation of last time this unit was updated (mer
+    :type last_updated: mongoengine.IntField
+    :ivar user_metadata: Bag of user supplied data to go along with this unit
+    :type user_metadata: mongoengine.DictField
+    :ivar storage_path: Location on disk where the content associated with this unit lives
+    :type storage_path: mongoengine.StringField
+
+    :ivar _ns: The namespace field (Deprecated), reading
+    :type _ns: mongoengine.StringField
+    :ivar unit_type_id: string representation of content unit type
+    :type unit_type_id: mongoengine.StringField
+    """
+
+    id = StringField(primary_key=True)
+    last_updated = IntField(db_field='_last_updated', required=True)
+    user_metadata = DictField(db_field='pulp_user_metadata')
+    storage_path = StringField(db_field='_storage_path')
+
+    # For backward compatibility
+    _ns = StringField(required=True)
+    unit_type_id = StringField(db_field='_content_type_id', required=True)
+    _source_location = None
+    _relative_path = None
+
+    meta = {
+        'abstract': True,
+    }
+
+    @classmethod
+    def post_init_signal(cls, sender, document):
+        """
+        The signal that is triggered before a unit is initialized
+
+        This is used to validate that the unit_key_fields attribute is set properly
+
+        :param sender: sender class
+        :type sender: object
+        :param document: Document that sent the signal
+        :type document: ContentUnit
+        """
+
+        if not hasattr(document, 'unit_key_fields'):
+            class_name = type(document).__name__
+            raise exceptions.PulpCodedException(error_codes.PLP0035, class_name=class_name)
+
+    @classmethod
+    def pre_save_signal(cls, sender, document, **kwargs):
+        """
+        The signal that is triggered before a unit is saved
+
+        :param sender: sender class
+        :type sender: object
+        :param document: Document that sent the signal
+        :type document: ContentUnit
+        """
+        if not document.id:
+            document.id = str(uuid.uuid4())
+        document.last_updated = dateutils.now_utc_timestamp()
```

Comment:
Does it make sense for this to only use standard library stuff? I'm trying to get us away from our litany of helper functions.